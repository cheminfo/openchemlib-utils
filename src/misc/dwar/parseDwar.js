import { ensureString } from 'ensure-string';

import { getParts } from './utils/getParts';
import { parseColumnbProperties } from './utils/parseColumnbProperties';
import { parseData } from './utils/parseData';

/**
 * Convert a DataWarrior database into a JSON object
 * @param {string} text
 * @returns
 */
export function parseDwar(text) {
  text = ensureString(text);
  const parts = getParts(text);
  improveParts(parts);
  return parts;
}

export function getCamelCase(name) {
  return name.replace(/[ -][a-z]/g, (string) => string[1].toUpperCase());
}

function improveParts(parts) {
  for (const key in parts) {
    switch (key) {
      case 'columnProperties':
        parts[key] = parseColumnbProperties(parts[key]);
        break;
      case 'data':
        break;
      default:
        parts[key] = parseDefault(parts[key]);
    }
  }
  const data = parseData(parts.data, {
    columnProperties: parts.columnProperties,
  });
  parts.data = data.entries;
  parts.rawData = data.rawEntries;
}

function parseDefault(lines) {
  const result = {};
  for (const line of lines) {
    const [key, value] = line.slice(1, -1).split('=');
    result[key] = value.slice(1, -1);
  }
  return result;
}
