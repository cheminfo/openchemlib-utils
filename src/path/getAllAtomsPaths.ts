import { Molecule } from 'openchemlib';

export interface GetAllAtomsPathsOptions {
  /**
   * Maximum path length to consider. All the paths from 0 to maxPathLength will be considered
   * @default 5
   */
  maxPathLength?: number;
}

interface AtomPath {
  path: number[];
  distance: number;
}

/**
 * We need to create an array of atoms
 * that contains an array of pathLength
 * that contains an array of object
 * @param molecule
 * @param options
 * @returns
 */
export function getAllAtomsPaths(
  molecule: Molecule,
  options: GetAllAtomsPathsOptions = {},
) {
  const { maxPathLength = 5 } = options;
  const allAtomsPaths = [];
  for (let i = 0; i < molecule.getAllAtoms(); i++) {
    const oneAtomPaths: AtomPath[][] = [];
    allAtomsPaths.push(oneAtomPaths);

    let atomPaths: AtomPath[] = [];
    atomPaths.push({ path: [i], distance: 0 });
    oneAtomPaths.push(atomPaths);

    let nextIndexes = [0];
    let nextAtoms = [i];

    for (let sphere = 1; sphere <= maxPathLength; sphere++) {
      atomPaths = [];
      oneAtomPaths.push(atomPaths);
      const currentIndexes = nextIndexes;
      const currentAtoms = nextAtoms;
      nextIndexes = [];
      nextAtoms = [];

      for (let i = 0; i < currentIndexes.length; i++) {
        const atom = currentAtoms[i];
        const index = currentIndexes[i];

        const previousPath = oneAtomPaths[sphere - 1][index].path;
        for (let conn = 0; conn < molecule.getConnAtoms(atom); conn++) {
          const connectedAtom = molecule.getConnAtom(atom, conn);
          if (previousPath.includes(connectedAtom)) continue;
          nextIndexes.push(atomPaths.length);
          nextAtoms.push(connectedAtom);
          atomPaths.push({
            path: [...previousPath, connectedAtom],
            distance: sphere,
          });
        }
      }
    }
  }
  return allAtomsPaths;
}
