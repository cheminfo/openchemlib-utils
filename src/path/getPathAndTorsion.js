/**
 * Calculates the path between 2 atoms
 * @param {import('openchemlib').Molecule} molecule
 * @param {number} from - index of the first atom
 * @param {number} to - index of the end atom
 * @param {number} maxLength - maximal length of the path
 */
export function getPathAndTorsion(molecule, from, to, maxLength) {
  let originalAtoms = []; // path before renumbering
  molecule.getPath(originalAtoms, from, to, maxLength + 1);
  let torsion;
  if (originalAtoms.length === 4) {
    torsion = molecule.calculateTorsion(originalAtoms);
  }

  return {
    atoms: originalAtoms,
    from,
    to,
    torsion,
    length: originalAtoms.length - 1,
  };
}
