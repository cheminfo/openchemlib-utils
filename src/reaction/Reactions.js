

export class Reactions {
  /**
   *
   * @param {object} [options={}]
   * @param {number} [options.maxDepth=5]
   * @param {function} [options.moleculeExtraInfo]
   * @param {boolean} [options.skipProcessed=true]
   */
  constructor(options = {}) {
    this.reactants = []; // array of array of molecules
    this.moleculeExtraInfo = options.moleculeExtraInfo;
    this.maxDepth = options.maxDepth ?? 5;
    this.skipProcessed = options.skipProcessed ?? true;
  }

  /**
   * @param {import('openchemlib').Molecule[]} reactants
   * @param {*} reactions
   */
  appendOneReactionReactants(reactants) {
    if (!Array.isArray(reactants)) {
      throw new TypeError('reactants must be an array');
    }
    this.reactants.push(reactants);
  }


  applyReactions(reactions, options = {}) {
    const { flattenProducts = true } = options;
  }

  filterTree(callback) { }

  filterProducts(callback) { }
}
