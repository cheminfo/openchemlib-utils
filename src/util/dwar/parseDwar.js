import { ensureString } from 'ensure-string';

import { getParts } from './utils/getParts.js';
import { parseColumnbProperties } from './utils/parseColumnbProperties.js';
import { parseData } from './utils/parseData.js';

/**
 * Convert a DataWarrior database into a JSON object
 * @param {string} text
 * @returns
 */
export function parseDwar(text) {
  text = ensureString(text);
  const parts = getParts(text);
  improveParts(parts);
  return parts;
}

function improveParts(parts) {
  for (const key in parts) {
    switch (key) {
      case 'columnProperties':
        parts[key] = parseColumnbProperties(parts[key]);
        break;
      case 'data':
        break;
      default:
        parts[key] = parseDefault(parts[key]);
    }
  }
  const data = parseData(parts.data, {
    columnProperties: parts.columnProperties,
  });
  parts.data = data.entries;
  parts.rawData = data.rawEntries;
}

function parseDefault(lines) {
  const result = {};
  for (const line of lines) {
    const [key, value] = line.slice(1, -1).split('=');
    result[key] = value.slice(1, -1);
  }
  return result;
}
