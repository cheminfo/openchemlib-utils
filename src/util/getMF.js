import atomSorter from 'atom-sorter';

/**
 * Calculate the molecular formula in 'chemcalc' notation taking into account fragments, isotopes and charges
 * {OCL.Molecule} [molecule] an instance of OCL.Molecule
 * @returns {object}
 */

export function getMF(molecule) {
  let entries = molecule.getFragments();
  let result = {};
  let parts = [];
  let allAtoms = [];

  entries.forEach((entry) => {
    let mf = getFragmentMF(entry, allAtoms);
    parts.push(mf);
  });

  let counts = {};
  for (let part of parts) {
    if (!counts[part]) counts[part] = 0;
    counts[part]++;
  }
  parts = [];
  for (let key of Object.keys(counts).sort()) {
    if (counts[key] > 1) {
      parts.push(counts[key] + key);
    } else {
      parts.push(key);
    }
  }

  result.parts = parts;
  result.mf = toMFString(allAtoms);
  return result;
}

function getFragmentMF(molecule, allAtoms) {
  let atoms = [];
  for (let i = 0; i < molecule.getAllAtoms(); i++) {
    let atom = {};
    atom.charge = molecule.getAtomCharge(i);
    atom.label = molecule.getAtomLabel(i);
    atom.mass = molecule.getAtomMass(i);
    atom.implicitHydrogens = molecule.getImplicitHydrogens(i);
    atoms.push(atom);
    allAtoms.push(atom);
  }
  return toMFString(atoms);
}

function toMFString(atoms) {
  let charge = 0;
  let mfs = {};
  for (let atom of atoms) {
    let label = atom.label;
    charge += atom.charge;
    if (atom.mass) {
      label = `[${atom.mass}${label}]`;
    }
    let mfAtom = mfs[label];
    if (!mfAtom) {
      mfs[label] = 0;
    }
    mfs[label] += 1;
    if (atom.implicitHydrogens) {
      if (!mfs.H) mfs.H = 0;
      mfs.H += atom.implicitHydrogens;
    }
  }

  let mf = '';
  let keys = Object.keys(mfs).sort(atomSorter);
  for (let key of keys) {
    mf += key;
    if (mfs[key] > 1) mf += mfs[key];
  }

  if (charge > 0) {
    mf += `(+${charge > 1 ? charge : ''})`;
  } else if (charge < 0) {
    mf += `(${charge < -1 ? charge : '-'})`;
  }
  return mf;
}
