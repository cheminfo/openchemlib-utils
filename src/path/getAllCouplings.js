import { tagAtom } from '../util/tagAtom';
import { makeRacemic } from '../util/makeRacemic';

/**
 * Returns an array of all the different atom diaIDs that are connected
 * {OCL}
 * {OCL.Molecule} molecule
 * {object} [options={}]
 * {string} [options.fromLabel='H']
 * {string} [options.toLabel='H']
 * {number} [options.minLength=1]
 * {number} [options.maxLength=4]
 */

function getAllCouplings(OCL, molecule, options = {}) {
  const {
    fromLabel = 'H',
    toLabel = 'H',
    minLength = 1,
    maxLength = 4,
  } = options;
  let paths = molecule.getAllPaths({
    fromLabel,
    toLabel,
    minLength,
    maxLength,
  });

  const minSphereSize = 0;
  const maxSphereSize = 2;

  let fragment = new OCL.Molecule(0, 0);
  for (let path of paths) {
    path.info = [];
    for (let fromTo of path.fromTo) {
      let atoms = [];
      molecule.getPath(atoms, fromTo[0], fromTo[1], path.pathLength);
      let torsion;
      if (atoms.length === 4) {
        torsion = molecule.calculateTorsion(atoms);
      }
      path.info.push({
        atoms,
        torsion,
      });

      if (!path.code) {
        path.code = [];

        let tmpMolecule = molecule.getCompactCopy();
        makeRacemic(OCL, tmpMolecule);
        tagAtom(OCL, tmpMolecule, atoms[0]);
        tagAtom(OCL, tmpMolecule, atoms[atoms.length - 1]);

        let atomMask = new Array(tmpMolecule.getAllAtoms()).fill(false);
        let atomList = [];
        let max = 0;
        let min = 0;
        for (let sphere = 0; sphere <= maxSphereSize; sphere++) {
          if (max === 0) {
            for (let atom of atoms) {
              atomMask[atom] = true;
              atomList.push(atom);
              max++;
            }
          } else {
            let newMax = max;
            for (let i = min; i < max; i++) {
              let atom = atomList[i];
              for (let j = 0; j < tmpMolecule.getConnAtoms(atom); j++) {
                let connAtom = tmpMolecule.getConnAtom(atom, j);
                atomMask[connAtom] = true;
                atomList[newMax++] = connAtom;
              }
            }
            min = max;
            max = newMax;
          }
          tmpMolecule.copyMoleculeByAtoms(fragment, atomMask, true, null);
          if (sphere >= minSphereSize) {
            path.code.push(
              fragment.getCanonizedIDCode(
                OCL.Molecule.CANONIZER_ENCODE_ATOM_CUSTOM_LABELS,
              ),
            );
          }
        }
      }
    }
  }
  return paths;
}

module.exports = getAllCouplings;
