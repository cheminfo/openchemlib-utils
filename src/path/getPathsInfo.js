import { getHoseCodesForPath } from '../hose/getHoseCodesForPath';
import { getAtomsInfo } from '../util/getAtomsInfo';
import { getConnectivityMatrix } from '../util/getConnectivityMatrix';

let fragment;

/**
 *
 * @param {OCL.Molecule} molecule
 * @param {object} [options={}]
 * @param {string} [opions.fromLabel='H']
 * @param {string} [opions.toLabel='H']
 * @param {string} [opions.minLength=1]
 * @param {string} [opions.maxLength=4]

 */
export function getPathsInfo(molecule, options = {}) {
  const {
    fromLabel = 'H',
    toLabel = 'H',
    minLength = 1,
    maxLength = 4,
  } = options;

  const OCL = molecule.getOCL();

  if (!fragment) {
    fragment = new OCL.Molecule(0, 0);
  }

  let fromAtomicNumber = OCL.Molecule.getAtomicNoFromLabel(fromLabel);
  let toAtomicNumber = OCL.Molecule.getAtomicNoFromLabel(toLabel);

  // we need to find all the atoms 'fromLabel' and 'toLabel'
  let atomsInfo = getAtomsInfo(molecule);

  let pathLengthMatrix = getConnectivityMatrix(molecule, {
    pathLength: true,
  });

  for (let from = 0; from < molecule.getAllAtoms(); from++) {
    atomsInfo[from].paths = [];
    for (let to = 0; to < molecule.getAllAtoms(); to++) {
      if (from !== to) {
        if (molecule.getAtomicNo(from) === fromAtomicNumber) {
          if (molecule.getAtomicNo(to) === toAtomicNumber) {
            let pathLength = pathLengthMatrix[from][to];
            if (pathLength >= minLength && pathLength <= maxLength) {
              atomsInfo[from].paths.push(
                getHoseCodesForPath(
                  molecule,
                  from,
                  to,
                  pathLength,
                  atomsInfo[to].oclID,
                ),
              );
            }
          }
        }
      }
    }
  }

  return atomsInfo;
}
