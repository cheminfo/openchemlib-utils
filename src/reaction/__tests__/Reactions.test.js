import { readFileSync } from 'fs';
import { join } from 'path';

import { MF } from 'mf-parser';
import { Molecule } from 'openchemlib';

import { parseDwar } from '../../misc/dwar/parseDwar.js';
import { getMF } from '../../util/getMF.js';
import { Reactions } from '../Reactions.js';

test('Reactions', () => {
  const dwar = readFileSync(
    join(__dirname, './data/ReactionMassFragmentation.dwar'),
    'utf8',
  );
  const database = parseDwar(dwar).data;

  console.log(database)
  const ionizations = database.filter((entry) => entry.Label === 'Ionization');
  const fragmentations = database.filter((entry) => entry.Label !== 'Ionization');

  const xtc = Molecule.fromSmiles('CC(CC1(=CC2(=C(C=C1)OCO2)))NC');

  const reactions = new Reactions({
    moleculeExtraInfo: (entry, molecule) => {
      const mf = getMF(molecule).mf;
      const mfInfo = new MF(mf).getInfo();
      entry.mf = mf;
      entry.mw = mfInfo.mass;
      entry.em = mfInfo.monoisotopicMass;
      entry.mz = mfInfo.observedMonoisotopicMass;
      entry.charge = mfInfo.charge;
    },
    maxDepth: 5,
    skipProcessed: true,
  });

  reactions.appendOneReactionReactants([xtc]);

  reactions.applyReactions(ionizations, { min: 1, max: 2, flattenProducts: true });

  reactions.applyReactions(fragmentations, { min: 1, max: 3, flattenProducts: true });

  reactions.filterTree((entry) => {
    console.log(reactions);
  });
});
