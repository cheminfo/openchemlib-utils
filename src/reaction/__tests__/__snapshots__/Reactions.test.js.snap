// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Reactions > mdma 1`] = `
Map {
  "dg~D@MBdie]v\\\\kahHBjh@@" => {
    "asProduct": undefined,
    "asReagent": false,
    "idCode": "dg~D@MBdie]v\\\\kahHBjh@@",
    "info": {
      "charge": 0,
      "em": 193.11027872702,
      "mf": "C11H15NO2",
      "mw": 193.24271923500675,
      "mz": undefined,
    },
  },
  "dg~DBMBmeJYW]gJxZB@jj@@" => {
    "asProduct": false,
    "asReagent": true,
    "idCode": "dg~DBMBmeJYW]gJxZB@jj@@",
    "info": {
      "charge": 1,
      "em": 194.11810375925,
      "mf": "C11H16NO2(+)",
      "mw": 194.2506599890625,
      "mz": 194.11755517934094,
    },
  },
  "dg~DBMB\`eJYW]gJxZB@jj@@" => {
    "asProduct": false,
    "asReagent": true,
    "idCode": "dg~DBMB\`eJYW]gJxZB@jj@@",
    "info": {
      "charge": 1,
      "em": 194.11810375925,
      "mf": "C11H16NO2(+)",
      "mw": 194.2506599890625,
      "mz": 194.11755517934094,
    },
  },
  "dg~DBMB\`eJYWugJxZB@jj@@" => {
    "asProduct": false,
    "asReagent": true,
    "idCode": "dg~DBMB\`eJYWugJxZB@jj@@",
    "info": {
      "charge": 1,
      "em": 194.11810375925,
      "mf": "C11H16NO2(+)",
      "mw": 194.2506599890625,
      "mz": 194.11755517934094,
    },
  },
  "dg~DBMBmyJYW]gJxZB@jj@@" => {
    "asProduct": false,
    "asReagent": true,
    "idCode": "dg~DBMBmyJYW]gJxZB@jj@@",
    "info": {
      "charge": -1,
      "em": 192.10245369479,
      "mf": "C11H14NO2(-)",
      "mw": 192.23477848095098,
      "mz": 192.10300227469907,
    },
  },
  "dclDB@iYRfUw\\\\kahHBj@@" => {
    "asProduct": true,
    "asReagent": true,
    "idCode": "dclDB@iYRfUw\\\\kahHBj@@",
    "info": {
      "charge": 1,
      "em": 163.07590459367,
      "mf": "C10H11O2(+)",
      "mw": 163.1935171106026,
      "mz": 163.07535601376094,
    },
  },
  "eF\`BL@" => {
    "asProduct": true,
    "asReagent": true,
    "idCode": "eF\`BL@",
    "info": {
      "charge": 0,
      "em": 31.04219916558,
      "mf": "CH5N",
      "mw": 31.057142878459935,
      "mz": undefined,
    },
  },
  "fHT\`@" => {
    "asProduct": true,
    "asReagent": true,
    "idCode": "fHT\`@",
    "info": {
      "charge": 1,
      "em": 15.02347509669,
      "mf": "CH3(+)",
      "mw": 15.03455815890258,
      "mz": 15.02292651678093,
    },
  },
  "dk^D@MBdie]vrnF\`\`Jj@@" => {
    "asProduct": true,
    "asReagent": true,
    "idCode": "dk^D@MBdie]vrnF\`\`Jj@@",
    "info": {
      "charge": 0,
      "em": 179.09462866255998,
      "mf": "C10H13NO2",
      "mw": 179.21610183015991,
      "mz": undefined,
    },
  },
}
`;

exports[`Reactions > mdma 2`] = `
[
  {
    "children": [
      {
        "depth": 1,
        "isValid": true,
        "molecules": [
          {
            "asProduct": false,
            "asReagent": true,
            "idCode": "dg~DBMBmeJYW]gJxZB@jj@@",
            "info": {
              "charge": 1,
              "em": 194.11810375925,
              "mf": "C11H16NO2(+)",
              "mw": 194.2506599890625,
              "mz": 194.11755517934094,
            },
          },
        ],
        "reaction": {
          "description": "",
          "doi": "https://doi.org/10.1186/1477-5956-10-S1-S8",
          "ionization": "[M+H]+",
          "kind": "ionization",
          "label": "Ionization",
          "reaction": "",
          "rxnCode": "gCh@AGj@vTTUHXECAdEzh\`!eF\`\`fLGtP#QP\` QX#!Ro\`BHoY^}GXk| !R_tp]Ngp",
          "section": "",
        },
      },
      {
        "depth": 1,
        "isValid": true,
        "molecules": [
          {
            "asProduct": false,
            "asReagent": true,
            "idCode": "dg~DBMB\`eJYW]gJxZB@jj@@",
            "info": {
              "charge": 1,
              "em": 194.11810375925,
              "mf": "C11H16NO2(+)",
              "mw": 194.2506599890625,
              "mz": 194.11755517934094,
            },
          },
        ],
        "reaction": {
          "description": "",
          "doi": "https://doi.org/10.1186/1477-5956-10-S1-S8",
          "ionization": "[M+H]+",
          "kind": "ionization",
          "label": "Ionization",
          "reaction": "",
          "rxnCode": "eMHBN\`vjMPp_yE{P!eFH\`fLGtP#QR QX#!R@BM_HqvJo@ !R_tp]Ngp",
          "section": "",
        },
      },
      {
        "depth": 1,
        "isValid": true,
        "molecules": [
          {
            "asProduct": false,
            "asReagent": true,
            "idCode": "dg~DBMB\`eJYWugJxZB@jj@@",
            "info": {
              "charge": 1,
              "em": 194.11810375925,
              "mf": "C11H16NO2(+)",
              "mw": 194.2506599890625,
              "mz": 194.11755517934094,
            },
          },
        ],
        "reaction": {
          "description": "",
          "doi": "https://doi.org/10.1186/1477-5956-10-S1-S8",
          "ionization": "[M+H]+",
          "kind": "ionization",
          "label": "Ionization",
          "reaction": "",
          "rxnCode": "eMHBN\`vjMPp_yE{P!eFH\`fLGtP#QR QX#!R@BM_HqvJo@ !R_tp]Ngp",
          "section": "",
        },
      },
    ],
    "depth": 0,
    "isValid": false,
    "molecules": [
      {
        "asProduct": undefined,
        "asReagent": false,
        "idCode": "dg~D@MBdie]v\\\\kahHBjh@@",
        "info": {
          "charge": 0,
          "em": 193.11027872702,
          "mf": "C11H15NO2",
          "mw": 193.24271923500675,
          "mz": undefined,
        },
      },
    ],
  },
]
`;
